feat: find SLSA provenance v0.2 published on npm registry

behnazh-w opened the pull request:
This PR adds a feature to detect SLSA provenance v0.2 for `npm` packages that are published on [npm registry](https://www.npmjs.com). An example SLSA provenance v0.2 for [sigstore/mock](https://www.npmjs.com/package/@sigstore/mock) package can be found [here](https://registry.npmjs.org/-/npm/v1/attestations/@sigstore/mock@0.1.0).

Some notes:

1. This PR does not add verification of these provenances. The verification will be handled separately: https://github.com/oracle/macaron/issues/550
2. Support for SLSA provenance v1 will be added in future: https://github.com/oracle/macaron/issues/549
3. A test for an `npm` package is added to the integration tests. If making REST Calls to the `npm` registry is not possible, developers can skip that test by setting `NO_NPM` environment variable before running `make integration-test`.

Closes #365

nathanwn commented on src/macaron/slsa_analyzer/checks/provenance_available_check.py:
```diff
@@ -195,7 +207,50 @@ def find_provenance_assets_on_package_registries(
+                                return []
+                            try:
+                                npm_provenance_payload = load_provenance_payload(download_path)
+                            except LoadIntotoAttestationError as loadintotoerror:
+                                logger.error("Error while loading provenance %s", loadintotoerror)
```
You probably also need to `return []` here, otherwise `npm_provenance_payload` can be unbound when you push it into `npm_info_entry_provenances`. This was reported by `pyright` on my side.
```suggestion
                            except LoadIntotoAttestationError as loadintotoerror:
                                logger.error("Error while loading provenance %s", loadintotoerror)
                                return []
```

tromai replied:
Pyright reports the same error for me.

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+
+        if not section.getboolean("enabled", fallback=True):
+            self.enabled = False
+            logger.debug("npm registry is disabled in section [{section_name}] of the .ini configuration file.")
```
You probably didn't mean to use f-string here.
```suggestion
            logger.debug(
                "npm registry is disabled in section [%s] of the .ini configuration file.",
                section_name,
            )
```

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        ----------
+        url: str
+            The attestation URL.
+        download_path: srt
```
```suggestion
        download_path: str
```

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        -------
+        bool
+            ``True`` if the asset is downloaded successfully; ``False`` if not.
+        """
```
I think you need to mention that this method raises `InvalidHTTPResponseError`.

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        """
+        try:
+            return self.npm_registry.download_attestation_payload(self.url, dest)
+        except InvalidPURLError as error:
```
Shouldn't this be `InvalidHTTPResponseError` instead?

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        * publish with "https://github.com/npm/attestation/tree/main/specs/publish/v0.1" predicateType
+        * SLSA with "https://slsa.dev/provenance/v0.2" predicateType
+
+        We download the SLSA provenance in this method.
```
I think this method, apart from downloading the attestation, also has some logic to validate the `predicateType` field and pick out the `dsseEnvelope` in the JSON payload. Could you also link to the example payload you showed in the PR description here?
```
https://registry.npmjs.org/-/npm/v1/attestations/@sigstore/mock@0.1.0
```
I think it will greatly help with understanding what this method is trying to do. Thanks, Behnaz.

tromai replied:
By `download the SLSA provenance`, I wonder whether we are mentioning about the whole attestation payload or we are mentioning about the attestation that has the  `https://slsa.dev/provenance/v0.2` predicateType?

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+    The API Documentation can be found here:
+    """
+
+    #: The namespace of the artifact on npm.
+    namespace: str | None
```
I think it may be helpful to have some explanation of what PURL `namespace` is supposed to be in the case of npm and why it can be `None`. I found the following sources:
* https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm. This says
> The namespace is used for the scope of a scoped NPM package.
* This page explains what a scoped NPM package is: https://docs.npmjs.com/cli/v10/using-npm/scope.

nathanwn commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+
+        Parameters
+        ----------
+        hostname : str
```
```suggestion
        hostname : str | None
```

nathanwn commented on src/macaron/slsa_analyzer/analyze_context.py:
```diff
@@ -91,25 +94,27 @@ def __init__(
         Returns
         -------
-        dict
+        dict[str : list[InTotoV01Statement]]
```
```suggestion
        dict[str, list[InTotoV01Statement | InTotoV1Statement]]
```

nathanwn commented on tests/slsa_analyzer/test_analyzer.py:
```diff
@@ -103,6 +106,41 @@ def test_resolve_analysis_target(
+    digest=st.text(),
+    available_domains=st.just(["github.com", "gitlab.com", "bitbucket.org"]),
+)
+def test_invalid_analysis_target(
+    purl_type: str,
+    namespace: str | None,
+    artifact_id: str,
+    version: str,
+    url: str,
+    branch: str,
+    digest: str,
+    available_domains: list[str],
+) -> None:
+    """Test the analysis target resolution with valid inputs."""
+    config = Configuration(
+        {
+            "purl": f"pkg:{purl_type}/{namespace}/{artifact_id}@{version}",
+            "path": url,
+            "branch": branch,
+            "digest": digest,
+        }
+    )
+    try:
+        Analyzer.to_analysis_target(config, available_domains)
+    except InvalidPURLError:
+        pass
```
I wonder what you are trying to test for here. Is it just making sure `Analyzer.to_analysis_target` does not crash unexpectedly with an unknown error type (apart from `InvalidPURLError`)?

behnazh-w replied:
It makes sure that random and invalid inputs are handled properly and either the `InvalidPURLError` exception is thrown or there is no other exception.

nathanwn commented on src/macaron/slsa_analyzer/analyzer.py:
```diff
@@ -540,6 +541,14 @@ def to_analysis_target(config: Configuration, available_domains: list[str]) -> A
             except ValueError as error:
                 raise InvalidPURLError(f"Invalid input PURL: {config.get_value('purl')}") from error
 
+        # Validate the purl object, which is user-controllable.
+        if parsed_purl:
+            purl_str = parsed_purl.to_string()
+            # The urlparse API sanitizes unsafe characters.
+            parsed_url = urllib.parse.urlparse(purl_str)
+            if purl_str != parsed_url.geturl():
+                raise ValueError("The PURL provided as input has unsafe characters.")
```
Because this seems important and probably can be tested, I wonder if we can extract this into a function and unit-test it.

tromai replied:
I agree.

tromai commented on src/macaron/slsa_analyzer/analyzer.py:
```diff
@@ -540,6 +541,14 @@ def to_analysis_target(config: Configuration, available_domains: list[str]) -> A
+        # Validate the purl object, which is user-controllable.
+        if parsed_purl:
+            purl_str = parsed_purl.to_string()
+            # The urlparse API sanitizes unsafe characters.
```
I think it's also good to document what kind of unsafe characters are we expecting it to have.

tromai commented on src/macaron/slsa_analyzer/analyzer.py:
```diff
@@ -540,6 +541,14 @@ def to_analysis_target(config: Configuration, available_domains: list[str]) -> A
 
+        # Validate the purl object, which is user-controllable.
+        if parsed_purl:
+            purl_str = parsed_purl.to_string()
```
If we are only trying to sanitize the original purl_string provided, do you think we should do it before trying to create a PackageURL object from that purl string?

tromai commented on tests/slsa_analyzer/test_analyzer.py:
```diff
@@ -103,6 +106,41 @@ def test_resolve_analysis_target(
+    digest: str,
+    available_domains: list[str],
+) -> None:
+    """Test the analysis target resolution with valid inputs."""
```
```suggestion
    """Test the analysis target resolution with valid inputs."""
```

nathanwn commented on tests/slsa_analyzer/checks/test_provenance_available_check.py:
```diff
@@ -50,6 +55,20 @@ def download_asset(self, url: str, download_path: str) -> bool:
 
 
+class MockNPMRegistry(NPMRegistry):
+    """Mocj NPMRegistry class."""
```
```suggestion
    """Mock NPMRegistry class."""
```

nathanwn commented on tests/slsa_analyzer/package_registry/test_npm_registry.py:
```diff
@@ -0,0 +1,131 @@
+def _create_npm_registry() -> NPMRegistry:
+    """Create an npm registry instance.
+
+    This private function is used as parameter for Hypothesis, which otherwise does not accept a fixture.
```
Hmm, I don't see this being used in any hypothesis tests.

behnazh-w replied:
Yes, forgot to remove it after moving the hypothesis test.

nathanwn commented on tests/slsa_analyzer/package_registry/test_npm_registry.py:
```diff
@@ -0,0 +1,131 @@
+        "namespace",
+        "artifact_id",
+        "version",
+        "expected",
```
Do you think `expected` is the best name here? It is not the expected value of any `assert`, but rather the path path in the URL. I'm thinking `package_path` or something similar may be a better name.

tromai replied:
I agree. Or if we want to keep `expected`, the value could change to the exact HTTPS URL that we finally compare too.

behnazh-w replied:
I prefer to the name `expected` as it is what the test expects to see as part of the `assert` statement.

nathanwn commented on src/macaron/slsa_analyzer/analyzer.py:
```diff
@@ -540,6 +541,14 @@ def to_analysis_target(config: Configuration, available_domains: list[str]) -> A
+            # The urlparse API sanitizes unsafe characters.
+            parsed_url = urllib.parse.urlparse(purl_str)
+            if purl_str != parsed_url.geturl():
+                raise ValueError("The PURL provided as input has unsafe characters.")
```
This case raises `ValueError`, but `ValueError` is neither documented in the method docstring nor caught at the call site of this method.
I also wonder if `ValueError` here should be converted to `InvalidPURLError` to be consistent with the rest of the method, or should it be a separate error type to be handled differently from the rest.

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        hostname: str | None = None,
+        attestation_endpoint: str | None = None,
+        request_timeout: int | None = None,
+        enabled: bool = True,
```
This parameter is different from `JfrogMavenRegistry` ([here](https://github.com/oracle/macaron/blob/52cecac033b384d69d3a711033e4b7f8fee49120/src/macaron/slsa_analyzer/package_registry/jfrog_maven_registry.py)) where the signature of enabled is: `enabled: bool | None = None` when I believe they have the same meaning. I wonder if there is any reason for this difference?

behnazh-w replied:
This PR adds support for `npm` which has its own requirements. Why do you bring up `JfrogMavenRegistry` in this review?

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        self.hostname = hostname or ""
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
```
When the instance of this class is created (before loading the values from `defaults.ini`) this attribute will be `True` and during the loading of values in `defaults.ini` it could be set to False if necessary. This approach is opposite to `JfrogMavenRegistry`. Is this desirable and should we keep it consistent both of them have the same `enabled` attribute?

behnazh-w replied:
Same answer. `npm` has its own requirements. I don't understand why `JfrogMavenRegistry` is mentioned here.

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+            ``True`` if the repo under analysis can be published to this package registry,
+            based on the given build tool.
+        """
+        compatible_build_tool_classes = [NPM, Yarn]
```
I think we should check if the `self.enabled` is False and return False here.

tromai commented on docs/source/pages/supported_technologies/index.rst:
```diff
@@ -86,8 +89,9 @@ Provenances
      -
-       * `SLSA provenance version 0.2 <https://slsa.dev/spec/v0.2/provenance>`_.
-       * The provenance should be published as a GitHub release asset
+       * | `SLSA provenance version 0.2 <https://slsa.dev/spec/v0.2/provenance>`_. The provenance should be
```
I think we should explicitly say that the provenance can `either` be  `published as a GitHub release asset.` or `published on the `npm registry <https://registry.npmjs.org>`_.`

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+                f'The "hostname" key is missing in section [{section_name}] of the .ini configuration file.'
+            )
+
+        self.attestation_endpoint = section.get("attestation_endpoint", "-/npm/v1/attestations")
```
I don't think we need to assign the default value of `-/npm/v1/attestations` as a fall back when the key `attestation_endpoint` is not available in `defaults.ini`. Because when en the user don't provide `attestation_endpoint` in their custom `defaults.ini`, for example:
```
[package_registry.npm]
# Set `enabled=False` to disable making REST API calls to the npm registry.
enabled = True
# npm registry host name.
hostname = registry.npmjs.org
# The attestation REST API.
request_timeout = 20
```
The final value of `self.attestation_endpoint` will be `-/npm/v1/attestations` as defined in the `src/macaron/config/defaults.ini` file.

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -0,0 +1,265 @@
+        if self.namespace:
+            path_params.append(self.namespace)
+        path_params.append(self.artifact_id)
+        path = f'{"/".join(path_params)}'
```
I think we can remove the f string here as `"/".join(path_params)` will return a string.

tromai commented on src/macaron/slsa_analyzer/checks/provenance_available_check.py:
```diff
@@ -478,7 +535,7 @@ def run_check(self, ctx: AnalyzeContext) -> CheckResultData:
             justification.append("Found provenance in release assets:")
             justification.extend(
-                [asset.name for asset in provenance_assets],
+                [asset.url for asset in provenance_assets],
```
I wonder  is the reason the asset's URL is displayed here instead of its artifact id, Behnaz?

nathanwn replied:
I assume with NPM attestation support added, displaying the URL to the attestation instead of the attestation name makes sense because NPM attestations are not files (like what we were dealing with previously) but HTTP response payload. `asset.name` here was always the name of the attestation file being downloaded but it is not applicable for the NPM attestation case.

I think URLs may even be the overall better output type here because they point directly to the place where the attestation can be found. Just the file name may not provide enough information, especially now that we have multiple places where an attestation can be discovered.

@behnazh-w Let us know if you are thinking the same, or if there is anything you want to add.

tromai reviewed (commented):
I have finished my first round of review.

behnazh-w commented on tests/slsa_analyzer/test_analyzer.py:
```diff
@@ -103,6 +106,41 @@ def test_resolve_analysis_target(
+    namespace=st.one_of(st.none(), st.text()),
+    artifact_id=st.text(),
+    version=st.text(),
+    url=st_pr.urls(),
```
It should be fine. If they make changes that break our test (which is unlikely), we will fix it. The  `urls()` strategy is pretty useful and I prefer to use it.

behnazh-w commented:
I have answered the questions and addressed comments in [79784f0](https://github.com/oracle/macaron/pull/551/commits/79784f05894629acba02a758c737e3c0d87be6c0).

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -96,14 +96,16 @@ def load_defaults(self) -> None:
 
     def is_detected(self, build_tool: BaseBuildTool) -> bool:
-        """Detect if artifacts of the repo under analysis can possibly be published to this package registry.
+        """Detect if artifacts under analysis can be published to this package registry.
```
Should this new py doc string for the `is_detected` method be updated in the base class `PackageRegistry` and its other children as well?

behnazh-w replied:
I think that is better to be done separately to keep this PR focused on `npm` support.

tromai commented on src/macaron/slsa_analyzer/package_registry/npm_registry.py:
```diff
@@ -128,20 +133,28 @@ def download_attestation_payload(self, url: str, download_path: str) -> bool:
 
         * publish with "https://github.com/npm/attestation/tree/main/specs/publish/v0.1" predicateType
         * SLSA with "https://slsa.dev/provenance/v0.2" predicateType
+        * SLSA with "https://slsa.dev/provenance/v1" predicateType
```
If the SLSA provenance v1 is added, I think we should change the above line to `Each npm package has three types of attestations`.

behnazh-w replied:
See commit [bae2fe0](https://github.com/oracle/macaron/pull/551/commits/bae2fe0376631a0a9081d8d507459bf5f14b6d5f).

tromai reviewed (changes requested):
I have reviewed all the changes related to my comments. There is only one new [comment](https://github.com/oracle/macaron/pull/551#discussion_r1402800276).

tromai reviewed (approved):
All of my comments have been resolved. The changes related to those LGTM.

nathanwn reviewed (approved):
All looks fine to me. Thanks for the PR.
